docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,416 @@
+* If multiple tasks are invoked simultaneously or in `pending` state, then they will be executed in a First-In-First-Out (FIFO) manner.
+
+> Note: Dependent ordering among tasks will be addressed later which may enable concurrent execution of tasks.
+
+#### Deletion
+
+Upon completion of the task, whether it has failed, succeeded, or been rejected, Etcd-druid will ensure the garbage collection of the custom resources and any other Kubernetes resources created to execute the task. This will be done according to the `.spec.ttlSecondsAfterFinished` if defined in the [spec](#spec), or a default expiry time will be assumed.
+
+### Use Cases
+
+#### Recovery from permanent quorum loss
+
+Recovery from permanent quorum loss involves two phases - identification and recovery - both of which are done manually today. This proposal intends to automate the latter. Recovery today is a [multi-step process](https://github.com/gardener/etcd-druid/blob/master/docs/operations/recovery-from-permanent-quorum-loss-in-etcd-cluster.md) and needs to be performed carefully by a human operator. Automating these steps would be prudent, to make it quicker and error-free. The identification of the permanent quorum loss would remain a manual process, requiring a human operator to investigate and confirm that there is indeed a permanent quorum loss with no possibility of auto-healing.
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should be a quorum loss in a multi-member etcd cluster. For a single-member etcd cluster, invoking this task is unnecessary as the restoration of the single member is automatically handled by the backup-restore process.
+* There should not already be a permanent-quorum-loss-recovery-task running for the same etcd cluster.
+
+#### Trigger on-demand snapshot compaction
+
+`Etcd-druid` provides a configurable [etcd-events-threshold](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/02-snapshot-compaction.md#druid-flags) flag. When this threshold is breached, then a [snapshot compaction](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/02-snapshot-compaction.md) is triggered for the etcd cluster. However, there are scenarios where an ad-hoc snapshot compaction may be required.
+
+##### Possible scenarios
+
+* If a human operator needs to verify the integrity of etcd cluster backups, particularly in cases of potential backup corruption or re-encryption, they can initiate an `on-demand snapshot compaction` task. The success or failure of this snapshot compaction can offer valuable insights into these scenarios.
```

Comment:
Didn't we decide that the reason for on-demand compaction is more towards creating a compacted full snapshot before performing a quorum loss recovery? So that restoration time is minimized. Checking for a corrupted backup should be its own Task, and should not be clubbed with on-demand compaction, which specifically talks about an on-demand compaction of full+delta snapshots into one single full snapshot, irrespective of what approach/implementation it uses underneath.

Reply:
> Didn't we decide that the reason for on-demand compaction is more towards creating a compacted full snapshot before performing a quorum loss recovery? 

yes, we did. I can add that too.

> Checking for a corrupted backup should be its own Task, and should not be clubbed with on-demand compaction, which specifically talks about an on-demand compaction of full+delta snapshots into one single full snapshot, irrespective of what approach/implementation it uses underneath.

why it need to be separate task doing the same thing ? If we need a separate task and that's the case then why we are introducing alerts for our compaction job ? 
AFAIR, we are introducing alerts for compaction feature, so that we can detect any cyber attack or snapshot corruption kind of scenarios quickly ... right ? I have mentioned the similar thing as `on-demand compaction` use case.

Reply:
The task defines the purpose for performing the operation. While snapshot compaction can help detect backup corruptions, it is not performed *in order to* detect backup corruptions. It's just a happy side-effect of the underlying implementation of snapshot compaction, that's all.

Alerts for snapshot compaction jobs tell us if a compaction job failed. When a human operator checks why the compaction job failed, the reason for it is then revealed, which could either be due to insufficient resources for the compaction job, error in the code, or backup corruption. The alert does not tell you if the backup is corrupted, but only tells you if the job failed. It's the human operator's job to decide how to use that information.

Similarly, the on-demand snapshot compaction task simply tells druid to perform an on-demand snapshot compaction, in order to compact the latest set of snapshots into a single full snapshot. If this task fails, then it's the human operator's job to check why the task failed, and might then realise that there was a backup corruption. But detection of backup corruption is not a primary goal of the task as such.

I would request you to mention for this point in the proposal: `as an added benefit, snapshot compaction also helps detect backup corruption`, rather than framing it like `if a human operator needs to verify the integrity of backups, they can use this...`

Reply:
> The task defines the purpose for performing the operation.

It comes under `possible scenarios` or you can say possible usage/application of given a task, but that's doesn't limit the purpose of the task, you can use it somewhere else also(other than mentioned in doc) if you wish and that's the whole purpose, to eventually provide a flexibility to invoke any `out-of-band` task for any scenario.

> While snapshot compaction can help detect backup corruptions, it is not performed in order to detect backup corruptions. It's just a happy side-effect of the underlying implementation of snapshot compaction, that's all.
Alerts for snapshot compaction jobs tell us if a compaction job failed. When a human operator checks why the compaction job failed, the reason for it is then revealed, which could either be due to insufficient resources for the compaction job, error in the code, or backup corruption. The alert does not tell you if the backup is corrupted, but only tells you if the job failed. It's the human operator's job to decide how to use that information.

yes, I completely agree with you but understand a scenario where operator got suspicious about integrity of backups and to confirm that they have to wait for this `compaction job` to get trigger(event threshold to reach), to avoid waiting, they can just trigger `on-demand snapshot compaction`, and that will provide a "valuable insights" (as mentioned in doc). In doc, we have never mentioned it will detect the corruption of backups but it can surely filter out the shoots(if succeed => no issue, if failed => then operator need to dig)

> Similarly, the on-demand snapshot compaction task simply tells druid to perform an on-demand snapshot compaction, in order to compact the latest set of snapshots into a single full snapshot. If this task fails, then it's the human operator's job to check why the task failed, and might then realise that there was a backup corruption. But detection of backup corruption is not a primary goal of the task as such.

you are correct and that's why we have explicitly mentioned this point in doc
```
The success or failure of this snapshot compaction can offer valuable insights into these scenarios.
```
We have never mentioned that this will detect the corruption case, it will simply provide some "insights" and filter out the domain of suspicious shoot and then operator need to find out as you already pointed out.

> as an added benefit, snapshot compaction also helps detect backup corruption, rather than framing it like if a human operator needs to verify the integrity of backups, they can use this...

this is already under the heading: `Possible Scenarios` then why to create one more sub heading as added benefit, IMO `added benefit` is seems to be just a synonym for `Possible Scenarios`.

Reply:
I have pushed the changes, PTAL